[SIMD] Intel support for splat

hyjorc1 opened the pull request:
#### 46b5fd3c0928a534c956bcda8820af4e97e9810c
<pre>
[SIMD] Intel support for splat
<a href="https://bugs.webkit.org/show_bug.cgi?id=248678">https://bugs.webkit.org/show_bug.cgi?id=248678</a>
rdar://102927968

Reviewed by Yusuke Suzuki.

Support WASM SIMD operation splat.
<a href="https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#constant">https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#constant</a>

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorSplat):
(JSC::MacroAssemblerX86_64::vectorSplatFloat32):
(JSC::MacroAssemblerX86_64::vectorSplatFloat64):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::pshufd):
(JSC::X86Assembler::pshufb):
(JSC::X86Assembler::pshuflw):
(JSC::X86Assembler::pshufhw):
(JSC::X86Assembler::X86InstructionFormatter::SingleInstructionBufferWriter::memoryModRM):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDSplat):

Canonical link: <a href="https://commits.webkit.org/257316@main">https://commits.webkit.org/257316@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7784850bb57dcd143e07ec4a1e1081d585c5a9e7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98576 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7786 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31700 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107998 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168262 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8296 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85164 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91110 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/105979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104258 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89841 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33278 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88105 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89371 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1719 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85137 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1632 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28730 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5037 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6570 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42158 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/87989 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3019 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19708 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2472,16 +2472,52 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        m_assembler.movq_rr(src, dest);
+        switch (lane) {
+        case SIMDLane::i64x2:
+            m_assembler.shufpd(dest, dest, 0);
```
Is `movddup` better? https://www.felixcloutier.com/x86/movddup

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2472,16 +2472,52 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+            return;
+        case SIMDLane::i16x8:
+            m_assembler.pshuflw(dest, dest, 0);
+            m_assembler.shufpd(dest, dest, 0);
```
`punpcklqdq` is better?

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2472,16 +2472,52 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        case SIMDLane::i8x16:
+            vectorReplaceLane(SIMDLane::i8x16, TrustedImm32(1), src, dest);
+            m_assembler.pshuflw(dest, dest, 0);
+            m_assembler.shufpd(dest, dest, 0);
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2472,16 +2472,52 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        case SIMDLane::f32x4:
+            if (src != dest)
+                m_assembler.movaps_rr(src, dest);
+            m_assembler.shufps(dest, dest, 0);
```
You can use `pshufd` if `src != dest`.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2472,16 +2472,52 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        case SIMDLane::f64x2:
+            if (src != dest)
+                m_assembler.movapd_rr(src, dest);
+            m_assembler.shufpd(dest, dest, 0);
```
`movddup` would be better.

justinmichaud commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2393,6 +2398,49 @@ class X86Assembler {
         m_formatter.immediate8((uint8_t) laneIndex);
     }
 
+    void pshufd(XMMRegisterID vd, XMMRegisterID vn, uint8_t controlBits)
```
Convention-wise, is this matching the manual? I thought intel usually put dest on the right. David, Yusuke?

Constellation replied:
Putting dest is the macro assembler's convention. In each arch's assembler, we rather follow each arch's assembly lang's convention. So align it to the other existing Intel assembler's method. Since Intel assembly
 lang uses dest on the left, we follow here.

But if the other method is using the dest on the right, do so.

Constellation replied:
Summary: follow to the X86Assembler class's convention.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2472,16 +2472,53 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+            return;
+        case SIMDLane::f64x2:
+            if (src != dest)
+                m_assembler.movapd_rr(src, dest);
```
Do we need this? I think we can always just do `m_assembler.movddup(dest, src);`

Constellation reviewed (approved):
r=me with comment. Please follow to the comment by Justin, checking what the right convention in X86Assembler, and follow to that.

hyjorc1 committed:
[SIMD] Intel support for splat
https://bugs.webkit.org/show_bug.cgi?id=248678
rdar://102927968

Reviewed by Yusuke Suzuki.

Support WASM SIMD operation splat.
https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#constant

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorSplat):
(JSC::MacroAssemblerX86_64::vectorSplatFloat32):
(JSC::MacroAssemblerX86_64::vectorSplatFloat64):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::pshufd):
(JSC::X86Assembler::pshufb):
(JSC::X86Assembler::pshuflw):
(JSC::X86Assembler::pshufhw):
(JSC::X86Assembler::X86InstructionFormatter::SingleInstructionBufferWriter::memoryModRM):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDSplat):

Canonical link: https://commits.webkit.org/257316@main
